File: src/ui/MainForm.java
package ui;

import java.awt.Dimension;
import java.awt.Font;
import java.awt.Toolkit;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.JButton;
import javax.swing.JFrame;
import javax.swing.JLabel;
import javax.swing.JMenu;
import javax.swing.JMenuBar;
import javax.swing.JMenuItem;
import javax.swing.JPanel;
import javax.swing.JTextField;

import driver.DataBaseDriver;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;


public class MainForm extends JFrame {
	private static final long serialVersionUID = -7109709456008554142L;

	private JPanel mainPanel;
	private JTextField textField;
	private static JMenuBar menuBar;
	private static DataBaseDriver dbDriver;

	/**
	 * Launch the application.
	 */
	public static void main(String[] args) {
		initialaze();
	}
	
	private static void initialaze(){
		MainForm frame = new MainForm();
		frame.setSize(465, 150);
		createMenu(frame);

		Dimension dimension = Toolkit.getDefaultToolkit().getScreenSize();
		int x = (int) ((dimension.getWidth() - frame.getWidth()) / 2);
		int y = (int) ((dimension.getHeight() - frame.getHeight()) / 2);
		frame.setLocation(x, y);
		frame.setVisible(true);
	}
	
	public static void create(DataBaseDriver dbUtils){
		dbDriver=dbUtils;
		initialaze();
	}
	

	private static void createMenu(JFrame frame) {
		menuBar = new JMenuBar();
		Font font = new Font("Verdana", Font.PLAIN, 11);
		JMenu fileMenu = new JMenu("Файл");
		fileMenu.setFont(font);

		JMenuItem addRecipe = new JMenuItem("Добавить рецепт");
		addRecipe.setFont(font);
		fileMenu.add(addRecipe);
		addRecipe.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				AddRecipeForm.create(dbDriver);
			}
		});
		
		JMenuItem myRecipe = new JMenuItem("Мои рецепты");
		myRecipe.setFont(font);
		myRecipe.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				MyRecipe.create(dbDriver);
			}
		});
		fileMenu.add(myRecipe);

		fileMenu.addSeparator();

		JMenuItem exitItem = new JMenuItem("Выход");
		exitItem.setFont(font);
		exitItem.addActionListener(new ActionListener() {
			public void actionPerformed(ActionEvent e) {
				System.exit(0);
			}
		});
		fileMenu.add(exitItem);

		menuBar.add(fileMenu);
		frame.setJMenuBar(menuBar);
	}

	/**
	 * Create the frame.
	 */
	public MainForm() {
		configureDefaultLayot();
	}

	private void configureDefaultLayot() {
		setResizable(false);
		setTitle("ЯВижуСыр");
		setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE);
		mainPanel = new JPanel();
		setContentPane(mainPanel);
		mainPanel.setLayout(null);

		JButton btnNewButton = new JButton("\u041F\u043E\u0434\u043E\u0431\u0440\u0430\u0442\u044C \u0440\u0435\u0446\u0435\u043F\u0442");
		btnNewButton.addMouseListener(new MouseAdapter() {
			@Override
			public void mouseClicked(MouseEvent e) {
				Result.create(dbDriver);
			}
		});
		btnNewButton.setBounds(120, 67, 216, 23);
		mainPanel.add(btnNewButton);

		textField = new JTextField();
		textField.setBounds(28, 36, 421, 20);
		mainPanel.add(textField);
		textField.setColumns(10);

		JLabel label = new JLabel("\u0421\u043F\u0438\u0441\u043E\u043A \u043F\u0440\u043E\u0434\u0443\u043A\u0442\u043E\u0432");
		label.setBounds(28, 11, 135, 14);
		mainPanel.add(label);
	}
}

File: src/by/bsuir/verpav/misoi/ui/MainForm.scala
package by.bsuir.verpav.misoi.ui

import java.awt.{Font, Toolkit, Image, FileDialog}
import java.awt.event.{ActionEvent, ActionListener}
import java.awt.image.BufferedImage
import java.io.File
import javax.imageio.ImageIO
import javax.swing._

import by.bsuir.verpav.misoi.harris.HarrisCornersDetector
import by.bsuir.verpav.misoi.neuronNetwork.NeuronNetwork
import scala.collection.mutable

/**
  * Created by Pavel_Verkhovtsov on 9/14/16.
  */
object MainForm extends JFrame with App{

  var baseImage: BufferedImage = _
  val imageLabel : JLabel = new JLabel()

  val neuronNetwork = new NeuronNetwork(20, 20)

  val sampleSets = mutable.Map[String, List[File]]()

  createMenu(this)
  createUIElements(this)
  setVisible(true)

  def setImageToLabel(): Unit = {
    val resizedImage = baseImage.getScaledInstance(imageLabel.getWidth, imageLabel.getHeight, Image.SCALE_SMOOTH)
    val icon = new ImageIcon(resizedImage)
    imageLabel.setIcon(icon)
    JOptionPane.showConfirmDialog(this,"Image updated", "Image updated", JOptionPane.CLOSED_OPTION)
  }

   private def loadImage(): Unit = {
    val chooser = new FileDialog(this, "Use a .png or .jpg extension", FileDialog.LOAD)
    chooser.setVisible(true)
    val filename = chooser.getDirectory + chooser.getFile
    baseImage = ImageIO.read(new File(filename))
    setImageToLabel()
  }

  private def loadSampleSetFolder(): Unit = {
    val chooser = new JFileChooser()
    chooser.setCurrentDirectory(new java.io.File("."))
    chooser.setFileSelectionMode(JFileChooser.DIRECTORIES_ONLY)
    chooser.setAcceptAllFileFilterUsed(false)
    if (chooser.showOpenDialog(this) == JFileChooser.APPROVE_OPTION) {
      val folderName = chooser.getSelectedFile.getAbsolutePath
      sampleSets.put(chooser.getSelectedFile.getName, new File(folderName).listFiles().toList)
    }
  }

  private def createUIElements(frame: JFrame): Unit = {
    setResizable(false)
    setTitle("Image processor")
    setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE)
    val mainPanel = new JPanel()
    setContentPane(mainPanel)
    mainPanel.setLayout(null)
    val dimension = Toolkit.getDefaultToolkit.getScreenSize
    setSize((dimension.getWidth / 2).toInt , (dimension.getHeight / 2).toInt)
    imageLabel.setSize((dimension.getWidth / 2).toInt, (dimension.getHeight / 2).toInt)
    imageLabel.setLocation(0, 0)
    this.add(imageLabel)
    val x1 = (dimension.getWidth - getWidth / 2).toInt
    val y1 = (dimension.getHeight - getHeight / 2).toInt
    setLocation(x1, y1)
  }

  def learn() : Unit = {
    neuronNetwork.learn(sampleSets)
    JOptionPane.showConfirmDialog(this,"Learning finished", "Learning finished", JOptionPane.CLOSED_OPTION)
  }

  def classify() : Unit = {
    val result = neuronNetwork.classify(baseImage)
    JOptionPane.showConfirmDialog(this, "Prognoses class : "+result, "Classification results", JOptionPane.CLOSED_OPTION)
  }

  def createMenu(frame: JFrame): Unit = {
    val menuBar = new JMenuBar()
    val font = new Font("Verdana", Font.PLAIN, 11)

    val fileMenu = new JMenu("File")
    fileMenu.setFont(font)

    val loadClass = new JMenuItem("Add class")
    loadClass.setFont(font)
    loadClass.addActionListener(loadSampleSetFolder _)
    fileMenu.add(loadClass)

    val loadItem = new JMenuItem("Load Image")
    loadItem.setFont(font)
    loadItem.addActionListener(loadImage _)
    fileMenu.add(loadItem)

    val exitItem = new JMenuItem("Exit")
    exitItem.setFont(font)
    exitItem.addActionListener(() => { frame.dispose(); System.exit(0)})
    fileMenu.add(exitItem)

    val clustering = new JMenu("Neurons network")
    clustering.setFont(font)

    val performClusteringStepItem = new JMenuItem("Learn")
    performClusteringStepItem.setFont(font)
    performClusteringStepItem.addActionListener(learn _)
    performClusteringStepItem.setAccelerator(KeyStroke.getKeyStroke('L', Toolkit.getDefaultToolkit.getMenuShortcutKeyMask))
    clustering.add(performClusteringStepItem)

    val performClusteringStepItemWithCustomParameters = new JMenuItem("Classify")
    performClusteringStepItemWithCustomParameters.setFont(font)
    performClusteringStepItemWithCustomParameters.addActionListener(classify _)
    performClusteringStepItemWithCustomParameters.setAccelerator(KeyStroke.getKeyStroke('S', Toolkit.getDefaultToolkit.getMenuShortcutKeyMask))
    clustering.add(performClusteringStepItemWithCustomParameters)

    menuBar.add(fileMenu)
    menuBar.add(clustering)
    frame.setJMenuBar(menuBar)
  }


  implicit def paramLambdaToActiveListener(lambda: (ActionEvent) => Unit): ActionListener = new ActionListener() {
    override def actionPerformed(e: ActionEvent): Unit = {
      lambda(e)
    }
  }

  implicit def lambdaToActiveListener(lambda: () => Unit): ActionListener = new ActionListener() {
    override def actionPerformed(e: ActionEvent): Unit = {
      lambda()
    }
  }
}